src/parser.cpp

```diff
@@ -8,62 +9,163 @@
+  // lowercase, and subtype is subtype, in ASCII lowercase.
   auto out = mimetype();
-  out.type = type;        // TODO: Convert this to ASCII lowercase
-  out.subtype = subtype;  // TODO: Convert this to ASCII lowercase
 
+  to_lower_ascii(type.data(), type.size());
+  to_lower_ascii(subtype.data(), subtype.size());
+
+  out.type = type;
+  out.subtype = subtype;
+
+  size_t position{0};
+
+  // 11. While position is not past the end of input:
+  while (position < input.size()) {
+    // Advance position by 1. (This skips past U+003B (;).)
+    position++;
+
+    // Collect a sequence of code points that are HTTP whitespace from input
+    // given position.
+    while (is_http_whitespace(input[position]) && position < input.size()) {
+      position++;
+    }
+
+    // Let parameterName be the result of collecting a sequence of code points
+    // that are not U+003B (;) or U+003D (=) from input, given position.
+    auto parameter_name_ending = input.find_first_of(";=", position);
+
+    std::string parameter_name{};
+
+    if (parameter_name_ending == std::string_view::npos) {
+      // If position is past the end of input, then break.
+      break;
+    } else {
+      // Optimization opportunity: Copy is only required if parameter_name is
+      // not lowercased.
+      parameter_name =
+          std::string(input.substr(position, parameter_name_ending - 1));
+      to_lower_ascii(parameter_name.data(), parameter_name.size());
+      position = parameter_name_ending;
+
+      // If the code point at position within input is U+003B (;), then
+      // continue.
+      if (input[position] == ';') continue;
+
+      // Advance position by 1. (This skips past U+003D (=).)
+      position++;
+    }
+
+    // Let parameterValue be null.
+    std::string parameter_value{};
+
+    // If the code point at position within input is U+0022 ("), then:
+    if (input[position] == '"') {
+      // Set parameterValue to the result of collecting an HTTP quoted string
+      // from input.
+      parameter_value = collect_http_quoted_string(input, position);
+
+      // Collect a sequence of code points that are not U+003B (;) from input,
+      // given position.
+      auto semicolon_index = input.find_first_of(';', position);
+
+      position = semicolon_index == std::string_view::npos ? input.size()
+                                                           : semicolon_index;
+    } else {
+      // Set parameterValue to the result of collecting a sequence of code
+      // points that are not U+003B (;) from input, given position.
+      auto semicolon_index = input.find_first_of(';', position);
+
+      if (semicolon_index == std::string_view::npos) {
+        semicolon_index = input.size();
+      }
+
+      std::string_view parameter_value_view =
+          input.substr(position, semicolon_index);
+
+      // Remove any trailing HTTP whitespace from parameterValue.
+      trim_http_whitespace(parameter_value_view);
```

Comment:
this will also remove leading whitespace

Reply:
You're right!